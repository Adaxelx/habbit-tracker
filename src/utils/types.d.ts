export type GridProps = {
  events: Event[];
  month: number;
  year: number;
  moveDate: Function;
};

export interface Event {
  _id: string;
  daysOfWeek: number[];
  title: string;
  timeStart?: string;
  timeEnd?: string;
  dateStart: string;
  dateEnd: string;
  label?: string | Label;
  description?: string;
  userId: string;
  checked?: boolean;
}

export interface EventLoop extends Event {
  numericStart: number;
  numericEnd: number;
}

export type Label = {
  _id: string;
  title: string;
  color: string;
};

export type CalendarTile = {
  day: number;
  events: EventLoop[];
  id: number;
};

export enum SIDES {
  LEFT = -1,
  RIGHT = 1,
}
