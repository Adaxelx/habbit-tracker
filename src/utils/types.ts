export type GridProps = {
  events: Event[];
  month: number;
  year: number;
  moveDate: Function;
  handleDayChange: Function;
};

export interface Event {
  _id: string;
  daysOfWeek: number[];
  title: string;
  timeStart?: string;
  timeEnd?: string;
  dateStart: string;
  dateEnd: string;
  label?: string | Label;
  description?: string;
  userId: string;
  checked?: [
    {
      day: number;
      month: number;
      year: number;
      _id: string;
    },
  ];
}

export interface EventLoop extends Event {
  numericStart: number;
  numericEnd: number;
}

export type Label = {
  _id: string;
  title: string;
  color: string;
  userId: string;
};

export type CalendarTile = {
  day: number;
  events: EventLoop[];
  id: number;
};

export enum SIDES {
  LEFT = -1,
  RIGHT = 1,
}

export type Color = { color?: string; id?: string };

export type NavProps = {
  header: string;
  moveDate: Function;
};

export interface DayCardProps {
  day: string;
  token: string | undefined;
}
