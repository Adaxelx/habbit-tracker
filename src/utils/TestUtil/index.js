export { default } from './TestUtil.ts';
