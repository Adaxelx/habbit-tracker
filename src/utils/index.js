export { default as TestUtil } from './TestUtil';
export { default as RenderWithRouter } from './RenderWithRouter';
export * from './validators';
export * from './types.d';
