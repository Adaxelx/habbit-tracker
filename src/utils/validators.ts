export const createRequiredObject = (name: string) => ({
  required: `Field ${name} is required.`,
});

export const validatePattern = (pattern: RegExp) => ({
  pattern: {
    value: pattern,
    message: 'Incorrect characters.',
  },
});

export const emailValidation = validatePattern(
  // eslint-disable-next-line no-control-regex
  /(?:[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*|"(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21\x23-\x5b\x5d-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])*")@(?:(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?|\[(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?|[a-z0-9-]*[a-z0-9]:(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21-\x5a\x53-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])+)\])/,
);

export const createRestrictedLengthObject = (name: string) => ({
  ...createRequiredObject(name),
  minLength: {
    value: 6,
    message: `Field ${name} must have at least 6 characters.`,
  },
  maxLength: {
    value: 30,
    message: `Field ${name} must have maximum 30 characters.`,
  },
});

const timeValidFuntion = (timeLess: string | undefined, time: string | undefined) => {
  const message = 'Your time is not in boundaries.';
  if (timeLess && time) {
    const [hours, minutes] = time.split(':');
    const [hoursMin, minutesMin] = timeLess.split(':');
    if (hoursMin < hours) {
      return true;
    }
    if (hoursMin === hours) {
      if (minutesMin < minutes) {
        return true;
      }
      return message;
    }
    return message;
  }
  return message;
};

export const createTimeValidation = (min: string | undefined, max: string | undefined) => ({
  validate: {
    min: (value: string) => timeValidFuntion(min, value),
    max: (value: string) => timeValidFuntion(value, max),
  },
});
