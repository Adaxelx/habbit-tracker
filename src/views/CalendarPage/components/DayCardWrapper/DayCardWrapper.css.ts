import styled from 'styled-components';

export const StyledWrapper = styled.div`
  display: flex;
  max-width: 100%;
  overflow-x: scroll;
`;
