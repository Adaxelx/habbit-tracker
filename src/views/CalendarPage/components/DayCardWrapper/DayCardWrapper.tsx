import React, { useLayoutEffect, useRef } from 'react';
import { DateTuple, Event, TokenType } from 'utils';
import { useCalendar, useQuery, useWindowSize } from 'hooks';
import { getEvents } from 'views/CalendarPage/CalendarPage.api';
import { weekDaysFull, getDayParsed, generateParsedDate } from 'constants/calendar';
import { StyledWrapper } from './DayCardWrapper.css';
import { DayCard } from '..';

interface DayCardWrapperProps {
  from: DateTuple;
  to: DateTuple;
  token: TokenType;
}

const DayCardWrapper = ({ from, to, token }: DayCardWrapperProps) => {
  const wrapper = useRef<HTMLDivElement>(null);
  const [width] = useWindowSize();

  const fromParsed = generateParsedDate(from);
  const toParsed = generateParsedDate(to);

  const [events] = useQuery<Event>([from, to, token], () => getEvents(token, fromParsed, toParsed));

  const [days] = useCalendar(events, from, to);

  useLayoutEffect(() => {
    if (days.length > 1) {
      if (wrapper && wrapper.current) {
        if (wrapper.current.scrollWidth > width) {
          const scroll = wrapper.current.scrollWidth / 2 - width / 2;
          wrapper?.current?.scroll(scroll, 0);
        }
      }
    }
  }, [days, wrapper]);

  const firstDayOfWeek = getDayParsed(new Date(fromParsed));
  let startDate = new Date(fromParsed).getTime();
  const today = new Date();
  today.setHours(0, 0, 0, 0);
  const todayTime = today.getTime();

  return (
    <StyledWrapper ref={wrapper}>
      {days.map(({ events: eventsArr, id }, i) => {
        const index = firstDayOfWeek + i > 6 ? (firstDayOfWeek + i) % 7 : firstDayOfWeek + i;
        const cardDate = new Date(startDate);
        cardDate.setHours(0, 0, 0, 0);
        const cardTime = cardDate.getTime();
        startDate += 24 * 60 * 60 * 1000;
        return (
          <DayCard
            active={todayTime === cardTime}
            key={id}
            header={weekDaysFull[index]}
            events={eventsArr}
          />
        );
      })}
    </StyledWrapper>
  );
};

export default DayCardWrapper;
