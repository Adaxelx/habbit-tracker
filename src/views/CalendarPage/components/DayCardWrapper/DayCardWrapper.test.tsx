import React from 'react';
import 'jest-styled-components';
import { TestUtil } from 'utils';
import DayCardWrapper from './DayCardWrapper';

describe('DayCardWrapper', () => {
  let util: TestUtil;
  let props = { from: [2021, 5, 4], to: [2021, 5, 10], token: 'abd' };
  beforeEach(() => {
    jest.clearAllMocks();
    util = new TestUtil(<DayCardWrapper {...props} />);
  });

  it('should match snapshot', () => {
    expect(util.render.asFragment()).toMatchSnapshot();
  });
});
