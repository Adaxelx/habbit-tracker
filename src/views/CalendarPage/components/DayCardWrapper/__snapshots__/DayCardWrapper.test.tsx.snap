// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DayCardWrapper should match snapshot 1`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  max-width: 100%;
  overflow-x: scroll;
}

.c1 {
  background-color: #FFFCAF;
  min-width: 275px;
  height: 500px;
  border: 1px solid #707070;
}

.c2 {
  font-size: 1.5rem;
  margin: 1.25rem 0;
  text-align: center;
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <h3
        class="c2"
      >
        Friday
      </h3>
    </div>
    <div
      class="c1"
    >
      <h3
        class="c2"
      >
        Saturday
      </h3>
    </div>
    <div
      class="c1"
    >
      <h3
        class="c2"
      >
        Sunday
      </h3>
    </div>
    <div
      class="c1"
    >
      <h3
        class="c2"
      >
        Monday
      </h3>
    </div>
    <div
      class="c1"
    >
      <h3
        class="c2"
      >
        Tuesday
      </h3>
    </div>
    <div
      class="c1"
    >
      <h3
        class="c2"
      >
        Wednesday
      </h3>
    </div>
    <div
      class="c1"
    >
      <h3
        class="c2"
      >
        Thursday
      </h3>
    </div>
  </div>
</DocumentFragment>
`;
